### DISCUSSION TITLE
Possibly exotic use case: Aliased relationships with ORM mapping trigger the mapper too early

### DISCUSSION THREAD

**Discussion by mcclurem** [sqlalchemy/sqlalchemy#12184] [Usage Questions]
I have a decently sized codebase where each model is broken into their own file with a __init__.py importing from each file
I have several objects that have a relationship to a time series of datapoints, and I want an explicit relationship for "most-recent-value from that series".
Best I can tell, this is supposed to be accomplished with aliased(). example of how I've been able to make this work is as follows:

device.py
```python
class Device(base):
    __tablename__ = "device"
    id: Mapped[int] = mapped_column(primary_key=True)
    most_recent_battery_data_point: Optional[BatteryDataPoint]  # forward annotation
```
battery_data_point.py:
```python
class BatteryDataPoint(Base):
    __tablename__ = "battery_soc"
    id: Mapped[int] = mapped_column(primary_key=True)
    device_id: Mapped[int] = mapped_column(ForeignKey("device.id"))
    timestamp: Mapped[datetime] = mapped_column(TIMESTAMP(timezone=True), index=True)
    soc: Mapped[float] = mapped_column(DECIMAL(precision=4, scale=1))

most_recent_battery_subquery = (
    select(BatteryDataPoint.device_id, func.max(BatteryDataPoint.timestamp).label("max_timestamp"))
    .group_by(BatteryDataPoint.device_id)
    .subquery("most_recent_battery_subquery")
)
latest_battery_query = select(BatteryDataPoint).select_from(
    Join(
        most_recent_battery_subquery,
        BatteryDataPoint,
        and_(
            BatteryDataPoint.device_id == most_recent_battery_subquery.c.device_id,
            BatteryDataPoint.timestamp == most_recent_battery_subquery.c.max_timestamp,
        ),
    )
)
```
__init__.py:
```python
from .battery_data_point import BatteryDataPoint, latest_battery_query
from .device import Device

Device.most_recent_battery_data_point = relationship(aliased(BatteryDataPoint, latest_battery_query.alias()), uselist=False, viewonly=True)  # type: ignore
```

If I put that aliased relationship directly into the device declaration, the mapper is triggered too early so I've only been able to use that pseudo forward-annotation. However, since the type of this forward annotation is Optional[BatteryDataPoint] it then causes an issue with mypy when I use it with joined load:
```python
db.session.query(m.Device).options(joinedload(m.Device.most_recent_battery_data_point))
```
`Argument 1 to "joinedload" has incompatible type "BatteryDataPoint | None"; expected "str | QueryableAttribute[Any]"`

Am I solving this problem incorrectly by doing this forward annotation with subsequent attachment of the real relationship?
If I /am/ doing the right thing, how should I be typing this forward annotation to make sure that it is both an `Optional[BatteryDataPoint]` as well as a `QueryableAttribute[Any]`? do I wrap it all up as `QueryableAttribute[Optional[BatteryDataPoint]]` ?

**Comment by zzzeek** [comment]
well you'd declare that like this, keep it `Mapped[]` since that's where your class-level behavior comes from:

```py

from typing import TYPE_CHECKING

class Device(base):
    __tablename__ = "device"
    id: Mapped[int] = mapped_column(primary_key=True)

    if TYPE_CHECKING:
        most_recent_battery_data_point: Mapped[Optional[BatteryDataPoint]]  # forward annotation
```

that said you can probably have your `most_recent_battery_data_point` inline using a callable as seen in [the second example here](https://docs.sqlalchemy.org/en/20/orm/join_conditions.html#integrating-aliasedclass-mappings-with-typing-and-avoiding-early-mapper-configuration), that's using a "before_mapper_configured" hook to keep everything together
